test/k8sT/Egress.go

```diff
@@ -149,6 +154,37 @@ var _ = SkipDescribeIf(func() bool {
 		// DNS query should work (pod-to-pod connectivity)
 		res = kubectl.ExecPodCmd(randomNamespace, srcPod, "dig kubernetes +time=2")
 		res.ExpectSuccess()
+
+		if ciliumOpts["tunnel"] == "disabled" {
+			// Reply traffic from pods that are selected by an egress policy should not
+			// be SNATed with the egress IP (only connections originating from these
+			// pods should go through egress gateway)
+			_, targetPodJSON := fetchPodsWithOffset(kubectl, randomNamespace, "server", testDS, hostIP, false, 1)
+
+			targetPodHostIP, err := targetPodJSON.Filter("{.status.hostIP}")
+			Expect(err).Should(BeNil(), "Cannot get target pod host IP")
+
+			targetPodIP, err := targetPodJSON.Filter("{.status.podIP}")
+			Expect(err).Should(BeNil(), "Cannot get target pod IP")
+
+			outsideNodeName, outsideNodeIP := kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+
+			// Add a route for the target pod's IP on the node running without Cilium to
+			// allow reaching it from outside the cluster
```

Comment:
Can't we expose the pod via a NodePort service instead? I'm guessing it's because we use a direct route that we can't run this in tunneling mode?

Reply:
> Can't we expose the pod via a NodePort service instead?

That's what I initially tried (https://github.com/cilium/cilium/commit/61eecd0123e548832348d27ec2ef801e0c6c893f) but that test was succeeding even without the reply traffic checks, since to my understanding:
* if the target pod is running on the same node to which we connect to, once we get a reply from the pod we will jump to a different program before even looking up the egress gw policy: https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L710-L714
* if the target pod is running on a different node, the node to which we connected to is going to SNAT the traffic with its own IP before forwarding it to the second node, and so even in this case the egress gw logic will be skipped since the reply traffic is destined to a remote node (https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L799)

> I'm guessing it's because we use a direct route that we can't run this in tunneling mode?

Yep, since we can't use a nodeport I run this test only in direct routing. This actually works even in tunneling mode (since `bpf_host` is attached anyway to the physical device), I'm just not sure it's worth having that test :thinking:

Reply:
Thanks for the explanation! It makes sense to me.

AFAICS, that means the additional check in the datapath is only necessary for native routing mode today. I'm tempted to limit the check to native routing, but... :point_down: 

> if the target pod is running on the same node to which we connect to, once we get a reply from the pod we will jump to a different program before even looking up the egress gw policy: https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L710-L714

This one relies on the assumption that we're using BPF to perform service resolution. If we ever enable the egress gateway to work with kube-proxy instead of KPR and we're relying on this assumption (e.g., we limited the check to only native routing mode), then the egress gateway may silently break. So probably best not to rely on that assumption.

Considering this, do you think it would be worth testing both pod-to-service in addition to direct pod-to-pod?

Reply:
(Whatever you decide, I think it would be worth a comment on the `ciliumOpts["tunnel"] == "disabled"` check to clarify the assumptions made here.)

Reply:
> Considering this, do you think it would be worth testing both pod-to-service in addition to direct pod-to-pod?

Yep, I guess it doesn't hurt to have an additional test for `external node -> nodeport service` since it can detect regressions on this front, so I'll add it back both for direct routing and tunneling :+1: